File: my-folder/2047-find-a-peak-element-ii/solution.cpp
class Solution {
public:
    vector<int> findPeakGrid(vector<vector<int>>& mat) {
        int low = 0;
        int high = mat[0].size();
        while(low<=high){
            int mid = (low + high) >> 1;
            int maxEle = -1;
            int maxInd = -1;
            for(int i = 0;i<mat.size();i++){
                if(mat[i][mid]>maxEle){
                    maxEle = mat[i][mid];
                    maxInd = i;
                }      
            }
            int left = mid-1>=0?mat[maxInd][mid-1]:-1;
            int right = mid+1<mat[0].size()?mat[maxInd][mid+1]:-1;
            if(mat[maxInd][mid]>left && mat[maxInd][mid]>right) return {maxInd,mid};
            else if(mat[maxInd][mid]<left) high = mid-1;
            else low = mid+1;

        }
        return {-1,-1};
        
    }
};

File: my-folder/2047-find-a-peak-element-ii/solution.java
class Solution {
    public int[] findPeakGrid(int[][] mat) {
        int left = 0;
        int right = mat[0].length-1;
        while(left<=right){
            int mid = (left+right)/2;
            
            int maxValue = -1;
            int maxInd = -1;
            for(int i=0;i<mat.length;i++){
                if(mat[i][mid]>maxValue){
                    maxValue = mat[i][mid];
                    maxInd = i;
                }
            }
            if((mid==0 || mat[maxInd][mid]>mat[maxInd][mid-1]) &&(mid==mat[0].length-1 || mat[maxInd][mid]>mat[maxInd][mid+1])) return (new int[] {maxInd,mid});
            else if(mid>0 && mat[maxInd][mid-1]>mat[maxInd][mid]) right = mid-1;
            else left = mid+1;
        }
        return (new int[] {-1,-1});
        
    }
}
